Add MaxResponseSizeMB to runtime config.

rohkhann opened the pull request:
## Why make this change?
This change adds in a new property to the host parameter of runtime config called MaxResponseSizeMB. This property will allow users to configure the amount of data that their host platform's memory can handle when streaming data from the underlying datasources.

## What is this change?
This change adds the maxresponseSizeMB property to the runtime config. In addition, it adds an accessor for that property which would be needed to implement the actual logic of streaming. 

## How was this tested?
Unit tests are added.

rohkhann commented:
/azp run

vadeveka commented on src/Config/ObjectModel/HostOptions.cs:
```diff
@@ -1,6 +1,45 @@
+{
+    /// <summary>
+    /// Dab engine can at maximum handle 187 MB of data in a single response from a source.
+    /// Json deserialization of a response into a string has a limit of 197020041 bytes which when converted to MB is 187 MB.
```
Where does 197020041 number come from? Limit in System.Text.Json dserializer ? Can we be more specific?

seantleonard replied:
Rohan, you can link to Microsoft Learn docs and to a section of code in .NET's github repo.

rohkhann replied:
Added.

seantleonard reviewed (changes requested):
When i have a dab-config.json where i set `"max-response-size-mb": null` and then run DAB's CLI -> `dab add Book --config "path/to/config.json" --source books --permissions "anonymous:create,read,update,delete" --graphql "book:books"`

The field `"max-response-size-mb"` disappears from my config. 
The use case is this: I am explicitly setting the property and value in my config. I am saying: "i want no limit to max response size, do NOT give me DAB's default value as that could change in the future. My config is explicitly telling dab what i want and the CLI shouldn't remove my explicitly defined field."

rohkhann commented on src/Service.Tests/Unittests/ConfigValidationUnitTests.cs:
```diff
@@ -2414,6 +2414,48 @@ public void TestRuntimeConfigSetupWithNonJsonConstructor()
+        /// <param name="expectedExceptionMessage">expected exception message in case there is exception.</param>
+        /// <param name="expectedMaxResponseSize">expected value in config.</param>
+        [DataTestMethod]
+        [DataRow(false, null, "", null,
+            DisplayName = "MaxDbResponseSizeMB should be null when no value provided in config.")]
+        [DataRow(false, 64, "", 64,
+            DisplayName = "Valid inputs of MaxResponseSize must be accepted and set in the config.")]
+        [DataRow(true, -1, $"{nameof(RuntimeConfig.Runtime.Host.MaxResponseSizeMB)} must be greater than 0 and <= 187 MB.", null,
+            DisplayName = "Valid inputs of MaxResponseSize must be accepted and set in the config.")]
+        [DataRow(true, 188, $"{nameof(RuntimeConfig.Runtime.Host.MaxResponseSizeMB)} must be greater than 0 and <= 187 MB.", null,
+            DisplayName = "Valid inputs of MaxResponseSize must be accepted and set in the config.")]
+        public void ValidateMaxResponseSizeInConfig(
+            bool exceptionExpected,
+            int? maxDbResponseSizeMB,
+            string expectedExceptionMessage,
+            int? expectedMaxResponseSize = null)
```
Hmm, I see similar practice for other test cases in dab :https://github.com/Azure/data-api-builder/blob/05588152e8c1c0a4ea86230c95b4460bf126b97f/src/Service.Tests/Unittests/ConfigValidationUnitTests.cs#L2190 and other similar test cases in ConfigValidationUnitTests.

seantleonard replied:
The displayNames must all be unique too. 

> Hmm, I see similar practice for other test cases in dab :

Sure, would those tests benefit from any changes to be more readable? In my opinion with this specific test, I'm merely aiming for improved readability. If not breaking up the test, perhaps order and consistent names of parameters make readability better?
- `providedMaxResponseSizeMB` next to `expectedMaxResponseSizeMB`
- proceeded by `exceptionExpected` and `expectedExceptionMessage`

rohkhann replied:
Fixed!

rohkhann commented:
> When i have a dab-config.json where i set `"max-response-size-mb": null` and then run DAB's CLI -> `dab add Book --config "path/to/config.json" --source books --permissions "anonymous:create,read,update,delete" --graphql "book:books"`
> 
> The field `"max-response-size-mb"` disappears from my config. The use case is this: I am explicitly setting the property and value in my config. I am saying: "i want no limit to max response size, do NOT give me DAB's default value as that could change in the future. My config is explicitly telling dab what i want and the CLI shouldn't remove my explicitly defined field."

Added similar flow to pagination options. Added all the things discussed in design doc.

rohkhann commented:
/azp run

seantleonard commented on src/Config/Converters/HostOptionsConverterFactory.cs:
```diff
@@ -0,0 +1,68 @@
+        public override void Write(Utf8JsonWriter writer, HostOptions value, JsonSerializerOptions options)
+        {
+            writer.WriteStartObject();
+            writer.WritePropertyName("cors");
+            JsonSerializer.Serialize(writer, value.Cors, options);
+            writer.WritePropertyName("authentication");
+            JsonSerializer.Serialize(writer, value.Authentication, options);
```
`HostOptions` method signature includes: `CorsOptions? Cors, AuthenticationOptions? Authentication`. Will this error out when `value.Cors` or `value.Authentication` is null?

rohkhann replied:
I validated that it would not by sending those in as null: 
![image](https://github.com/Azure/data-api-builder/assets/124841904/75e6c8ae-f8e4-45ff-a6af-234e102fb82e)
![image](https://github.com/Azure/data-api-builder/assets/124841904/1b74e352-8406-4a49-a7bb-744d8112914f)

seantleonard commented on src/Config/ObjectModel/HostOptions.cs:
```diff
@@ -1,6 +1,77 @@
+    /// <summary>
+    /// Flag which informs CLI and JSON serializer whether to write MaxResponseSizeMB.
+    /// property and value to the runtime config file.
+    /// When user doesn't provide the MaxResponseSizeMB property/value, which signals DAB to use the default,
```
Should clarify that `UserProvidedMaxResponseSizeMB` is true only when a user provides a non-null value. Dab will ignore the property even if the user writes the following in their config:
```json
{
    "max-response-size-mb": null
}
```

rohkhann replied:
Ok let me add that in.

rohkhann replied:
Added in.

rohkhann commented:
/azp run

seantleonard commented on src/Config/ObjectModel/HostOptions.cs:
```diff
@@ -1,6 +1,78 @@
+    /// <summary>
+    /// Dab engine can at maximum handle 158 MB of data in a single response from a source.
+    /// Json deserialization of a response into a string has a limit of 166,666,666 bytes which when converted to MB is 158 MB.
+    /// ref: enforcing code: https://github.com/dotnet/runtime/blob/main/src/libraries/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.cs#L103
+    /// ref: Json constant: https://github.com/stephentoub/runtime/blob/main/src/libraries/System.Text.Json/src/System/Text/Json/JsonConstants.cs
```
nit: https://github.com/dotnet/runtime link is for latest runtime code, you'd want to ensure you select tag for .net8 and/or .net6 to demonstrate the code responsible for behavior we see today.

nit: https://github.com/stephentoub is .net engineer's personal repo `System.Text.Json/src/System/Text/Json/JsonConstants.cs`.. any chance you can make this direct link to dotnet/runtime repo?

rohkhann replied:
Yes let me fix this

rohkhann replied:
fixed.

seantleonard reviewed (approved):
last two nits on comments. Thanks for making these changes!

rohkhann commented:
/azp run
